revocation/revocation.go

```diff
@@ -0,0 +1,122 @@
+//
+// TODO: add CRL support
+func (r *revocation) Validate(certChain []*x509.Certificate, signingTime time.Time) error {
+	return r.OCSPStatus(certChain, signingTime)
+	// TODO: add CRL support
+	// https://github.com/notaryproject/notation-core-go/issues/125
+}
+
+// OCSPStatus checks OCSP, returns nil if all certs in the chain are not revoked.
+// If there is an error, it will return one of the errors defined in the revocation/ocsp package in errors.go.
+// (e.g. if a certificate in the chain is revoked by OCSP and there are no other errors, it will return ocsp.RevokedError)
+func (r *revocation) OCSPStatus(certChain []*x509.Certificate, signingTime time.Time) error {
+	return ocsp.OCSPStatus(ocsp.Options{
+		CertChain:       certChain,
+		SigningTime:     signingTime,
+		HTTPClient:      r.httpClient,
+		MergeErrorsFunc: r.mergeErrorsFunc,
+	})
+}
+
+// SetMergeErrorsFunction allows you to specify an alternative function to merge errors if the default does not fit your use case. You can also pass nil to reset it to the DefaultMergeErrors function
+func (r *revocation) SetMergeErrorsFunction(mergeErrorsFunc func(errors []error) error) {
+	if mergeErrorsFunc == nil {
+		r.mergeErrorsFunc = DefaultMergeErrors
+	} else {
+		r.mergeErrorsFunc = mergeErrorsFunc
+	}
+}
+
+// DefaultMergeErrors condenses errors for a list of errors (either for cert chain or OCSP servers) into one primary error
+func DefaultMergeErrors(errorList []error) error {
+	var result error
+	if len(errorList) > 0 {
+		result = errorList[0]
+
+		for _, err := range errorList {
+			if err == nil {
+				continue
+			}
+			switch t := err.(type) {
+			case ocsp.RevokedError:
+				// There is a revoked certificate
+				// return since any cert being revoked means leaf is revoked
+				return t
+			case ocsp.CheckOCSPError:
+				// There is an error checking
+				// return since any cert having error means chain has error (return earliest)
+				return t
+			case ocsp.UnknownStatusError:
+				// A cert in the chain has status unknown
+				// will not return immediately (in case one is revoked or has error), but will override other chain errors
+				result = t
+			case ocsp.PKIXNoCheckError:
+				// A cert in the chain only had server responses without the pkix nocheck extension
+				// will not return immediately (in case there is higher level error)
+				// will override TimeoutError, NoOCSPServerError, and nil, but not UnknownStatusError
+				if _, ok := result.(ocsp.UnknownStatusError); !ok || result == nil {
+					result = t
+				}
+			case ocsp.TimeoutError:
+				// A cert in the chain timed out while checking OCSP
+				// will not return immediately (in case there is higher level error)
+				// will override nil and NoOCSPServerError, but not UnknownInOCSPError or PKIXNoCheckError
+				if _, ok := result.(ocsp.NoOCSPServerError); ok || result == nil {
+					result = t
```

Comment:
Should timeout and CheckOCSPError have same priority ?

Reply:
Also in my opinion we should move this logic to notation-go i.e. return error for each cert in cert chain.

Reply:
I'd say timeout and CheckOCSPError should be different based on their cause/permanence. If you retry a verification that had a timeout error in an hour (or maybe even with an adjusted timeout threshold), there's a chance it succeeds. However, if you retry a verification that had a CheckOCSPError, more often than not there is some other issue with the response/request/chain/cert that will need to be addressed first before it can be resolved. That is why I prioritized it higher.

w.r.t. moving this logic to notation-go, what are your thoughts on having the status/validate function return the entire list of errors, then have the merging function separated? That would still allow ease of use to users to prevent them from having to parse it, but would also allow provide granularity and insight into each error in the cert chain so that we can log them in notation-go (essentially in notation-go we could log each error, then determine the overarching status based on the merge of errors)

Reply:
Referring to @shizhMSFT's suggestion [here](https://github.com/notaryproject/notation-core-go/pull/134#discussion_r1151422211), this merge function could instead just return Revoked, Unknown, or OK, allowing a user to provide the error list and get a simple result. If they want to parse this differently than our default behavior, they would have the error list to do so on their end (essentially eliminating the need for the SetMergeErrorsFunction).